url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/engine/Range.java

```diff
@@ -0,0 +1,203 @@
+        return hc;
+    }
+
+    public boolean equals(Object obj) {
+        if (!(obj instanceof Range))
+            return false;
+        Range r = (Range) obj;
+        if (!Reflect.equals(r.namespace, namespace))
+            return false;
+        if (!Reflect.equals(r.dim1, dim1))
+            return false;
+        if (!Reflect.equals(r.dim2, dim2))
+            return false;
+        return true;
+    }
+
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        if (namespace != null) {
+            sb.append(namespace);
+            sb.append('!');
+        }
+
+        if (dim1Name != null) {
+            sb.append(dim1Name);
+        } else {
+            sb.append(dim1);
+        }
+
+        if (dim2Name != null) {
+            sb.append(':');
+            sb.append(dim2Name);
+        } else if (dim2 != null) {
+            sb.append(':');
+            sb.append(dim2);
+        }
+
+        return sb.toString();
+    }
+
+    public static Range valueOf(String var) throws ExprException {
+        if (var == null) {
+            throw new ExprException("Invalid empty variable name");
+        }
+        String namespace = null;
+        int ni = var.indexOf('!');
+        if (ni != -1) {
+            namespace = var.substring(0, ni);
+            if (!namespace.startsWith("'") &&
+                    !GridReference.isValidVariable(namespace)) {
+                throw new ExprException("Invalid namespace: " + namespace);
+            }
+            var = var.substring(ni + 1);
+        }
+
+        String dim1Name = null;
+        GridReference dim1 = null;
+        String dim2Name = null;
+        GridReference dim2 = null;
+
+        int di = var.indexOf(':');
+        if (di != -1) {
+            dim1Name = var.substring(0, di);
+            dim1 = GridReference.valueOf(dim1Name);
+            if (dim1 == null && !GridReference.isValidVariable(dim1Name)) {
+                throw new ExprException("Invalid variable: " + dim1Name);
+            }
+            dim2Name = var.substring(di + 1);
+            dim2 = GridReference.valueOf(dim2Name);
+            if (dim2 == null && !GridReference.isValidVariable(dim2Name)) {
+                throw new ExprException("Invalid variable: " + dim2Name);
+            }
+        } else {
+            dim1Name = var;
+            dim1 = GridReference.valueOf(var);
+            if (dim1 == null && !GridReference.isValidVariable(dim1Name)) {
+                throw new ExprException("Invalid variable: " + dim1Name);
+            }
+
+        }
+
+        return new Range(namespace, dim1Name, dim1, dim2Name, dim2);
+    }
+
+    public static Range toRange(ExprArray array, GridReference offset) {
+        if (offset == null) {
+            offset = new GridReference(1, 1);
+        }
+
+        GridReference dim2 = new GridReference(offset.getColumn() +
+                array.columns() - 1, offset.getRow() + array.rows() - 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;array&#39; (index: 0 | type: ExprArray). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;offset&#39;. The same parameter type is validated here for example: url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/engine/GridMap.java:47. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).